primitives/src/merkle_tree/internal.rs

```diff
@@ -226,6 +231,92 @@ where
+where
+    E: Element,
+    H: DigestAlgorithm<E, I, T>,
+    I: Index + From<u64>,
+    Arity: Unsigned,
+    T: NodeValue,
+{
+    let leaves: Vec<_> = elems.into_iter().collect();
+    let num_leaves = leaves.len() as u64;
+    let capacity = num_traits::checked_pow(Arity::to_u64(), height).ok_or(
+        PrimitivesError::ParameterError("Merkle tree size too large.".to_string()),
+    )?;
+
+    if num_leaves > capacity {
+        Err(PrimitivesError::ParameterError(
+            "Too many data for merkle tree".to_string(),
+        ))
+    } else if num_leaves > 0 {
+        let mut cur_nodes = leaves
+            .into_iter()
+            .enumerate()
+            .chunks(Arity::to_usize())
+            .into_iter()
+            .map(|chunk| {
+                let children = chunk
+                    .map(|(pos, elem)| {
+                        if (pos as u64) < num_leaves - 1 {
+                            Box::new(MerkleNode::ForgettenSubtree {
+                                value: H::digest_leaf(&I::from(pos as u64), elem.borrow()),
+                            })
+                        } else {
+                            let pos = I::from(pos as u64);
+                            Box::new(MerkleNode::Leaf {
+                                value: H::digest_leaf(&pos, elem.borrow()),
+                                pos,
+                                elem: elem.borrow().clone(),
+                            })
+                        }
+                    })
+                    .pad_using(Arity::to_usize(), |_| Box::new(MerkleNode::Empty))
+                    .collect_vec();
+                Box::new(MerkleNode::<E, I, T>::Branch {
+                    value: digest_branch::<E, H, I, T>(&children),
+                    children,
+                })
+            })
+            .collect_vec();
+        for i in 1..cur_nodes.len() - 1 {
+            cur_nodes[i] = Box::new(MerkleNode::ForgettenSubtree {
+                value: cur_nodes[i].value(),
+            })
+        }
+        for _ in 1..height {
+            cur_nodes = cur_nodes
+                .into_iter()
+                .chunks(Arity::to_usize())
+                .into_iter()
+                .map(|chunk| {
+                    let children = chunk
+                        .pad_using(Arity::to_usize(), |_| {
+                            Box::new(MerkleNode::<E, I, T>::Empty)
+                        })
+                        .collect_vec();
+                    Box::new(MerkleNode::<E, I, T>::Branch {
+                        value: digest_branch::<E, H, I, T>(&children),
+                        children,
+                    })
+                })
+                .collect_vec();
+            for i in 1..cur_nodes.len() - 1 {
+                cur_nodes[i] = Box::new(MerkleNode::ForgettenSubtree {
+                    value: cur_nodes[i].value(),
+                })
+            }
+        }
+        Ok((cur_nodes[0].clone(), num_leaves))
```

Comment:
What's the difference between `cur_nodes[0]` and `cur_nodes[0].clone()` here?

Reply:
`cur_nodes` is a local variable and it cannot be moved outside.

Reply:
Ah right, it's a local vector; we can't just move the first element of it outside.